kong/types.go

```diff
@@ -267,3 +270,84 @@ type Role struct {
+	Actions   []*string         `json:"actions,omitempty" yaml:"actions,omitempty"`
+	Negative  *bool             `json:"negative,omitempty" yaml:"negative,omitempty"`
+	Role      *PermissionRoleID `json:"role,omitempty" yaml:"role,omitempty"`
+	Comment   *string           `json:"comment,omitempty" yaml:"comment,omitempty"`
+}
+
+// MarshalJSON marshals an endpoint permission into a suitable form for the Kong admin API
+func (e *EndpointPermission) MarshalJSON() ([]byte, error) {
+	type ep struct {
+		CreatedAt *int              `json:"created_at,omitempty" yaml:"created_at,omitempty"`
+		Workspace *string           `json:"workspace,omitempty" yaml:"workspace,omitempty"`
+		Endpoint  *string           `json:"endpoint,omitempty" yaml:"endpoint,omitempty"`
+		Actions   *string           `json:"actions,omitempty" yaml:"actions,omitempty"`
+		Negative  *bool             `json:"negative,omitempty" yaml:"negative,omitempty"`
+		Role      *PermissionRoleID `json:"role,omitempty" yaml:"role,omitempty"`
+		Comment   *string           `json:"comment,omitempty" yaml:"comment,omitempty"`
+	}
+	var actions []string
+	for _, action := range e.Actions {
+		actions = append(actions, *action)
+	}
+	return json.Marshal(&ep{
+		CreatedAt: e.CreatedAt,
+		Workspace: e.Workspace,
+		Endpoint:  e.Endpoint,
+		Actions:   String(strings.Join(actions, ",")),
+		Comment:   e.Comment,
+	})
+}
+
+// EntityPermission represents an RBAC Entity Permission in Kong Enterprise
+type EntityPermission struct {
+	CreatedAt  *int              `json:"created_at,omitempty" yaml:"created_at,omitempty"`
+	EntityID   *string           `json:"entity_id,omitempty" yaml:"entity_id,omitempty"`
+	EntityType *string           `json:"entity_type,omitempty" yaml:"entity_type,omitempty"`
+	Actions    []*string         `json:"actions,omitempty" yaml:"actions,omitempty"`
+	Negative   *bool             `json:"negative,omitempty" yaml:"negative,omitempty"`
+	Role       *PermissionRoleID `json:"role,omitempty" yaml:"role,omitempty"`
+	Comment    *string           `json:"comment,omitempty" yaml:"comment,omitempty"`
+}
+
+// MarshalJSON marshals an endpoint permission into a suitable form for the Kong admin API
+func (e *EntityPermission) MarshalJSON() ([]byte, error) {
+	type ep struct {
+		CreatedAt  *int              `json:"created_at,omitempty" yaml:"created_at,omitempty"`
+		EntityID   *string           `json:"entity_id,omitempty" yaml:"entity_id,omitempty"`
+		EntityType *string           `json:"entity_type,omitempty" yaml:"entity_type,omitempty"`
+		Actions    *string           `json:"actions,omitempty" yaml:"actions,omitempty"`
+		Negative   *bool             `json:"negative,omitempty" yaml:"negative,omitempty"`
+		Role       *PermissionRoleID `json:"role,omitempty" yaml:"role,omitempty"`
+		Comment    *string           `json:"comment,omitempty" yaml:"comment,omitempty"`
+	}
+	var actions []string
+	for _, action := range e.Actions {
+		actions = append(actions, *action)
+	}
```

Comment:
Question: Why is this required? Why can't we mirror what is present in Kong's Admin API?

Reply:
Unfortunately the way that Kong's Admin API accepts POSTs and PUTs is in a different form (comma-separated vs. actual array) than what you receive from a GET request.